tests/test_dir_traversal.py

```diff
@@ -95,8 +95,8 @@ class DirectoryTraversalVector(Generic[T]):
 
     id: str
     given: List[BaseEntry]
-    add: List[Tuple[str, Optional[T]]] = field(default_factory=lambda: [(".", None)])
-    ignore: List[str] = field(default_factory=list)
+    add: List[Tuple[str, Tuple[T, ...]]] = field(default_factory=lambda: [(".", ())])
```

Comment:
Agree that these are not sufficiently (self-)documented. What we're setting up here is a list of calls to `traversal.add(...)`. Each item corresponds to one `.add()` call, and consists of:
- a path, relative to `tmp_path` (in the test below, we do `tmp_path / path` to turn it into an absolute path, and that allows us to use a `str` here, which cuts down on boilerplate, i.e. not having to wrap every path in `Path(...)`)
- a collection of arbitrary data to _attach_. This can be _any_ type, and the contract is that while traversing underneath the added path, this data will be provided back to the caller via the `.attached` member on the `TraversalStep` object.

I am thinking of adding a `NamedTuple` like this:

```diff
+class AddCall(NamedTuple):
+    path: Union[Path, str]
+    attach: Tuple[T, ...] = ()
+
+
 @dataclass
 class DirectoryTraversalVector(Generic[T]):
     """Test vectors for DirectoryTraversal."""
 
     id: str
     given: List[BaseEntry]
-    add: List[Tuple[str, Tuple[T, ...]]] = field(default_factory=lambda: [(".", ())])
+    add: List[AddCall] = field(default_factory=lambda: [AddCall(".")])
     skip_dirs: List[str] = field(default_factory=list)
     expect: List[ExpectedTraverseStep] = field(default_factory=list)
     expect_alternatives: Optional[List[List[ExpectedTraverseStep]]] = None
```

And each vector would then be changed like this:
```diff
     DirectoryTraversalVector(
         "one_subdir_plus__attach_data_on_both",
         given=[Dir("sub")],
-        add=[(".", (123,)), ("sub", (456,))],
+        add=[AddCall(".", (123,)), AddCall("sub", (456,))],
         expect=[
             ExpectedTraverseStep(".", subdirs=["sub"], attached=[123]),
             ExpectedTraverseStep("sub", attached=[123, 456]),
         ],
     ),
```

or even more verbose:
```diff
     DirectoryTraversalVector(
         "one_subdir_plus__attach_data_on_both",
         given=[Dir("sub")],
-        add=[(".", (123,)), ("sub", (456,))],
+        add=[
+            AddCall(path=".", attach=(123,)),
+            AddCall(path="sub", attach=(456,)),
+        ],
         expect=[
             ExpectedTraverseStep(".", subdirs=["sub"], attached=[123]),
             ExpectedTraverseStep("sub", attached=[123, 456]),
         ],
     ),
```

What do you prefer? Is this sufficient, or should we do something more/different?

Reply:
I think the last option is the most readable. Indeed, it adds verbosity to the test examples, but on the other hand one knows what is being tested at a glance.

Reply:
Fixed in [13f9373](https://github.com/tweag/FawltyDeps/pull/387/commits/13f9373ef6b2f90c610c38e409e9314611baca36)